.circleci/config.yml

```diff
@@ -1,56 +1,378 @@
-      #     SA_PASSWORD: "Admin12345"
-      #     ACCEPT_EULA: "Y"
-      # - image: cockroachdb/cockroach-unstable:latest
-      #   command: start --insecure
+  sqlite:
+    docker:
+      - image: circleci/node:10.15.0 # sqlite doesn't require another image
+  sqljs:
+    docker:
+      - image: circleci/node:10.15.0 # sqljs doesn't require another image
 
+commands:
+  prepare-for-tests:
+    parameters:
+      node_folder:
+        type: enum
+        enum:
+          - node8
+          - node10
+          - node12
     steps:
-      - checkout
+      - attach_workspace:
+          at: ~/typeorm/
       - run: sudo npm install -g npm@latest
       - run: cp ormconfig.circleci.json ormconfig.json
 
       # Download and cache dependencies
       - restore_cache:
           keys:
-          - v1-dependencies-{{ checksum "package.json" }}
-          # fallback to using the latest cache if no exact match is found
-          - v1-dependencies-
+            - << parameters.node_folder >>-v1-dependencies-{{ checksum "package.json" }}
+            # fallback to using the latest cache if no exact match is found
+            - << parameters.node_folder >>-v1-dependencies-
 
       - run: npm install
 
       - save_cache:
           paths:
             - node_modules
-          key: v1-dependencies-{{ checksum "package.json" }}
-
-      # - run: docker-compose up -d
+          key: << parameters.node_folder >>-v1-dependencies-{{ checksum "package.json" }}
 
       # run tests
       - run: npm run lint
+
+  run-db-tests:
+    parameters:
+      database:
+        type: enum
+        enum:
+          - mariadb
+          - mongodb
+          - mssql
+          - mysql
+          - oracle
+          - postgres
+          - sqlite
+          - sqljs
+    steps:
+      - attach_workspace:
+          at: ~/typeorm/
+      - run: cp ormconfigs/ormconfig.<< parameters.database >>.json ormconfig.json
       - run: npm test
+
+jobs:
+  checkout_code:
+    docker:
+      - image: circleci/node:10.15.0
+    working_directory: ~/typeorm/checkout/
+    steps:
+      - checkout
+      - run:
+          name: Copy source for node versions
+          command: |
+            mkdir ~/typeorm/node8/
+            mkdir ~/typeorm/node10/
+            mkdir ~/typeorm/node12/
+
+            cp -R ~/typeorm/checkout/. ~/typeorm/node8/
+            cp -R ~/typeorm/checkout/. ~/typeorm/node10/
+            cp -R ~/typeorm/checkout/. ~/typeorm/node12/
+      - persist_to_workspace:
+          root: ~/typeorm/
+          paths:
+            - ./checkout/
+            - ./node8/
+            - ./node10/
+            - ./node12
+
+  test-node8: &test-template
+    working_directory: ~/typeorm/node8/
+    docker:
+      - image: circleci/node:8.16.1
+    steps:
+      - prepare-for-tests:
+          node_folder: node8
+      - persist_to_workspace:
+          root: ~/typeorm/
+          paths:
+            - ./node8/
+
+  test-node8-mariadb: &mariadb-template
```

Comment:
This step actually runs tests. In this case, it runs the MariaDB tests on Node 8. Once again, we see `&mariadb-template`, which allows the Node 10 & 12 MariaDB tests to only specify values which are different. Unlike the last step, this one does actually reduce duplicate configuration since the only property with node specific values is `working_directory`.

Each of the test running steps will specify:
* which executor to use (so that the appropriate DB container us used)
* the working directory to use (which depends on the node version)
* the steps to run (which is almost always just `run-db-tests) with which DB is being tested, so we can copy the ormconfig for that DB.